File: src/main/java/tool/dbcleanup/DBCleanupTool.java
package tool.dbcleanup;

import core.db.DBManager;
import core.file.hrf.HRF;
import core.gui.RefreshManager;
import core.model.HOVerwaltung;
import core.model.match.MatchKurzInfo;
import core.model.enums.MatchType;
import core.util.HODateTime;
import core.util.HOLogger;

import java.sql.Timestamp;
import java.time.temporal.ChronoUnit;
import java.util.Calendar;
import java.util.GregorianCalendar;

import javax.swing.JFrame;


/**
 * HO Database Cleanup Tool
 * Removes old HRFs and old matches from the DB to speedup HO
 * 
 * @author flattermann <HO@flattermann.net>
 */
public class DBCleanupTool {
	public static int REMOVE_NONE = -1;
	public static int REMOVE_ALL = 0;

	public DBCleanupTool() {
		
	}
	
	public void showDialog(JFrame owner) {
		new DBCleanupDialog(owner, this);
	}

	/**
	 * Remove old HRFs
	 * 
	 * @param keepWeeks remove HRFs older than x weeks (-1=keep All, 0=remove All)
	 * @param autoRemove if true, automatically remove all HRFs except the first per training week  
	 */
	public void cleanupHRFs (int keepWeeks, boolean autoRemove) {
		Timestamp removeDate = null;
		if (keepWeeks >= 0) {
			var hrfDateFrom = HODateTime.now().minus(keepWeeks*7, ChronoUnit.DAYS);
			removeDate = hrfDateFrom.toDbTimestamp();
		}
		if (removeDate != null || autoRemove) {
			cleanupHRFs(removeDate, autoRemove);
		}
	}
	
	/**
	 * Remove old HRFs
	 * 
	 * @param removeDate remove HRFs older than x (null=keep all)
	 * @param autoRemove if true, automatically remove all HRFs except the first per training week  
	 */
	public void cleanupHRFs (Timestamp removeDate, boolean autoRemove) {
		HOLogger.instance().debug(getClass(),
				"Removing old HRFs: removeDate=" + removeDate + ", autoRemove=" + autoRemove);
		HRF[] allHrfs = DBManager.instance().loadAllHRFs( true);
		var latestHrf = DBManager.instance().getLatestHRF();
		int lastSeason = -1;
		int lastWeek = -1;
		int counter = 0;
		for (HRF curHrf : allHrfs) {
			int curId = curHrf.getHrfId();
			var curDate = curHrf.getDatum();
			var htWeek = curDate.toTrainingWeek();
			int curHtSeasonTraining = htWeek.season;
			int curHtWeekTraining = htWeek.week;
			boolean remove = false;
			if (removeDate != null && removeDate.after(curDate.toDbTimestamp())) {
				remove = true;
			} else if (autoRemove) {
				if (lastSeason == curHtSeasonTraining && lastWeek == curHtWeekTraining) {
					remove = true;
				} else {
					lastSeason = curHtSeasonTraining;
					lastWeek = curHtWeekTraining;
				}
			}
			// Do not remove the latest HRF
			if (remove && curId != latestHrf.getHrfId()) {
				HOLogger.instance().debug(getClass(),
						"Removing Hrf: " + curId + " @ " + curDate + " (" + curHtSeasonTraining + "/" + curHtWeekTraining + ")");
				DBManager.instance().deleteHRF(curId);
				counter++;
			}
		}
		HOLogger.instance().debug(getClass(), "Removed " + counter + "/" + allHrfs.length + " HRFs from DB!");
		if (counter > 0) {
			reInitHO();
		}
	}
	
	/**
	 * Remove old matches from DB (by weeks)
	 * 
	 * @param keepWeeksOwnMatches	keep x weeks of own matches (-1=keep All, 0=remove All)
	 * @param keepWeeksOwnFriendlies keep x weeks of own friendlies (-1=keep All, 0=remove All)
	 * @param keepWeeksOtherMatches keep x weeks of other matches (-1=keep All, 0=remove All)
	 * @param keepWeeksOtherFriendlies keep x weeks of other friendlies (-1=keep All, 0=remove All)
	 */
	void cleanupMatches (int keepWeeksOwnMatches, int keepWeeksOwnFriendlies, int keepWeeksOtherMatches, int keepWeeksOtherFriendlies) {
		Timestamp removeDateOwnMatches = null;
		Timestamp removeDateOwnFriendlies = null;
		Timestamp removeDateOtherMatches = null;
		Timestamp removeDateOtherFriendlies = null;
		if (keepWeeksOwnMatches >= 0) {
			Calendar cal = new GregorianCalendar();
			cal.add(Calendar.WEEK_OF_YEAR, -keepWeeksOwnMatches);
			removeDateOwnMatches = new Timestamp(cal.getTimeInMillis());
		}
		if (keepWeeksOwnFriendlies >= 0) {
			Calendar cal = new GregorianCalendar();
			cal.add(Calendar.WEEK_OF_YEAR, -keepWeeksOwnFriendlies);
			removeDateOwnFriendlies = new Timestamp(cal.getTimeInMillis());
		}
		if (keepWeeksOtherMatches >= 0) {
			Calendar cal = new GregorianCalendar();
			cal.add(Calendar.WEEK_OF_YEAR, -keepWeeksOtherMatches);
			removeDateOtherMatches = new Timestamp(cal.getTimeInMillis());
		}
		if (keepWeeksOtherFriendlies >= 0) {
			Calendar cal = new GregorianCalendar();
			cal.add(Calendar.WEEK_OF_YEAR, -keepWeeksOtherFriendlies);
			removeDateOtherFriendlies = new Timestamp(cal.getTimeInMillis());
		}
		if (removeDateOwnMatches != null || removeDateOwnFriendlies != null
				|| removeDateOtherMatches != null || removeDateOtherFriendlies != null) {
			cleanupMatches (removeDateOwnMatches, removeDateOwnFriendlies, removeDateOtherMatches, removeDateOtherFriendlies);
		}
	}
	
	/**
	 * Remove old matches from DB (by date)
	 * 
	 * @param removeDateOwnMatches	remove own matches older than x (null=keep All)
	 * @param removeDateOwnFriendlies	remove own friendlies older than x (null=keep All)
	 * @param removeDateOtherMatches remove other matches older than x (null=keep All)
	 * @param removeDateOtherFriendlies remove other friendlies older than x (null=keep All)
	 */
	private void cleanupMatches (Timestamp removeDateOwnMatches, Timestamp removeDateOwnFriendlies, 
			Timestamp removeDateOtherMatches, Timestamp removeDateOtherFriendlies) {
		HOLogger.instance().debug(getClass(),  
				"Removing old matches: "
				+ "removeDateOwnMatches="+removeDateOwnMatches 
				+ ", removeDateOwnFriendlies="+removeDateOwnFriendlies 
				+ ", removeDateOtherMatches="+removeDateOtherMatches
				+ ", removeDateOtherFriendlies="+removeDateOtherFriendlies);
		var kurzInfos = DBManager.instance().getMatchesKurzInfo(-1);
		int counter = 0;
		int myTeamId = HOVerwaltung.instance().getModel().getBasics().getTeamId();
		for (MatchKurzInfo curKurzInfo : kurzInfos) {
			Timestamp curMatchDate = curKurzInfo.getMatchSchedule().toDbTimestamp();
			int curMatchId = curKurzInfo.getMatchID();
			MatchType curMatchType = curKurzInfo.getMatchType();
			boolean isMyMatch = (curKurzInfo.getHomeTeamID() == myTeamId || curKurzInfo.getGuestTeamID() == myTeamId);
			boolean removeMatch = false;
			if (isMyMatch) {
				if (curMatchType.isFriendly()) {
					if (removeDateOwnFriendlies != null && removeDateOwnFriendlies.after(curMatchDate)) {
						// Remove friendly of my team
						removeMatch = true;
					}
				} else if (removeDateOwnMatches != null && removeDateOwnMatches.after(curMatchDate)) {
					// Remove non-friendly of my team
					removeMatch = true;					
				}
			} else {
				if (curMatchType.isFriendly()) {
					if (removeDateOtherFriendlies != null && removeDateOtherFriendlies.after(curMatchDate)) {
						// Remove friendly of other team
						removeMatch = true;
					}
				} else if (removeDateOtherMatches != null && removeDateOtherMatches.after(curMatchDate)) {
					// Remove non-friendly of other team
					removeMatch = true;					
				}
				
			}
			if (removeMatch) {
				// Remove match
				HOLogger.instance().debug(getClass(), 
						"Removing match "+curMatchId+" @ "+curMatchDate+ " (myMatch="+isMyMatch+", type="+curMatchType+")");
	            DBManager.instance().deleteMatch(curKurzInfo);
				counter++;
			} 
		}
		HOLogger.instance().debug(getClass(), "Removed " + counter + "/" + kurzInfos.size() + " matches from DB!");
		if (counter > 0) {
			reInitHO();
		}
	}
	
	private void reInitHO() {
        RefreshManager.instance().doReInit();
	}
}

File: src/main/java/tool/dbcleanup/DBCleanupTool.kt
package tool.dbcleanup

import core.db.DBManager
import core.gui.RefreshManager
import core.model.HOVerwaltung
import core.model.enums.MatchType
import core.model.match.MatchKurzInfo
import core.util.HODateTime
import core.util.HOLogger
import java.sql.Timestamp
import java.time.temporal.ChronoUnit
import java.util.*
import javax.swing.JFrame

/**
 * HO Database Cleanup Tool
 * Removes old HRFs and old matches from the DB to speedup HO
 *
 * @author flattermann <HO@flattermann.net>
 */
class DBCleanupTool {

    companion object {
        var REMOVE_NONE: Int = -1
        var REMOVE_ALL: Int = 0
    }

    fun showDialog(owner: JFrame?) {
        DBCleanupDialog(owner, this)
    }

    /**
     * Remove old HRFs
     *
     * @param keepWeeks remove HRFs older than x weeks (-1=keep All, 0=remove All)
     * @param autoRemove if true, automatically remove all HRFs except the first per training week
     */
    fun cleanupHRFs(keepWeeks: Int, autoRemove: Boolean) {
        var removeDate: Timestamp? = null
        if (keepWeeks >= 0) {
            val hrfDateFrom = HODateTime.now().minus(keepWeeks * 7, ChronoUnit.DAYS)
            removeDate = hrfDateFrom.toDbTimestamp()
        }
        if (removeDate != null || autoRemove) {
            cleanupHRFs(removeDate, autoRemove)
        }
    }

    /**
     * Remove old HRFs
     *
     * @param removeDate remove HRFs older than x (null=keep all)
     * @param autoRemove if true, automatically remove all HRFs except the first per training week
     */
    private fun cleanupHRFs(removeDate: Timestamp?, autoRemove: Boolean) {
        HOLogger.instance().debug(
            javaClass,
            "Removing old HRFs: removeDate=$removeDate, autoRemove=$autoRemove"
        )
        val allHrfs = DBManager.instance().loadAllHRFs(true)
        val latestHrf = DBManager.instance().latestHRF
        var lastSeason = -1
        var lastWeek = -1
        var counter = 0

        for (curHrf in allHrfs) {
            val curId = curHrf.hrfId
            val curDate = curHrf.datum
            val htWeek = curDate.toTrainingWeek()
            val curHtSeasonTraining = htWeek.season
            val curHtWeekTraining = htWeek.week
            var remove = false
            if (removeDate != null && removeDate.after(curDate.toDbTimestamp())) {
                remove = true
            } else if (autoRemove) {
                if (lastSeason == curHtSeasonTraining && lastWeek == curHtWeekTraining) {
                    remove = true
                } else {
                    lastSeason = curHtSeasonTraining
                    lastWeek = curHtWeekTraining
                }
            }
            // Do not remove the latest HRF
            if (remove && curId != latestHrf.hrfId) {
                HOLogger.instance().debug(
                    javaClass,
                    "Removing Hrf: $curId @ $curDate ($curHtSeasonTraining/$curHtWeekTraining)"
                )
                DBManager.instance().deleteHRF(curId)
                counter++
            }
        }
        HOLogger.instance().debug(javaClass, "Removed $counter/${allHrfs.size} HRFs from DB!")
        if (counter > 0) {
            reInitHO()
        }
    }

    private fun calculateDateLimit(numWeeks: Int): Timestamp {
        val cal: Calendar = GregorianCalendar()
        cal.add(Calendar.WEEK_OF_YEAR, -numWeeks)
        return Timestamp(cal.timeInMillis)
    }

    /**
     * Remove old matches from DB (by date)
     *
     * @param cleanupDetails Parameters for cleanup
     */
    fun cleanupMatches(cleanupDetails: CleanupDetails) {
        HOLogger.instance().debug(
            javaClass,
            "Removing old matches: ownTeamMatchTypes=${cleanupDetails.ownTeamMatchTypes}, " +
                    "ownTeamWeeks=${cleanupDetails.ownTeamWeeks}, " +
                    "otherTeamMatchTypes=${cleanupDetails.otherTeamMatchTypes}, " +
                    "otherTeamWeeks=${cleanupDetails.otherTeamWeeks}"
        )

        var counter = 0
        val kurzInfos = DBManager.instance().getMatchesKurzInfo(-1)
        val myTeamId = HOVerwaltung.instance().model.basics.teamId

        for (curKurzInfo in kurzInfos) {
            val curMatchId = curKurzInfo.matchID
            var removeMatch = false

            if (checkDeleteMatch(myTeamId, curKurzInfo, cleanupDetails)) {
                HOLogger.instance().info(
                    javaClass,
                    "Match to be deleted: ${curKurzInfo.matchID}, " +
                            "matchType=${curKurzInfo.matchType}, matchDate=${curKurzInfo.matchSchedule}"
                )
                removeMatch = true
            }

            if (removeMatch) {
                // Remove match
                HOLogger.instance().debug(
                    javaClass,
                    "Removing match $curMatchId"
                )
                DBManager.instance().deleteMatch(curKurzInfo)
                counter++
            }
        }
        HOLogger.instance().debug(javaClass, "Removed $counter/${kurzInfos.size} matches from DB!")

        if (counter > 0) {
            reInitHO()
        }
    }

    private fun checkDeleteMatch(
        myTeamId: Int,
        curKurzInfo: MatchKurzInfo,
        cleanupDetails: CleanupDetails
    ): Boolean {
        val curMatchDate = curKurzInfo.matchSchedule.toDbTimestamp()
        val curMatchType = curKurzInfo.matchType
        val isMyMatch = (curKurzInfo.homeTeamID == myTeamId || curKurzInfo.guestTeamID == myTeamId)

        val ownMatchToDelete = (isMyMatch && cleanupDetails.ownTeamMatchTypes.contains(curMatchType)
                && calculateDateLimit(cleanupDetails.ownTeamWeeks).after(curMatchDate))
        val otherMatchToDelete = (!isMyMatch && cleanupDetails.otherTeamMatchTypes.contains(curMatchType)
                && calculateDateLimit(cleanupDetails.otherTeamWeeks).after(curMatchDate))

        return ownMatchToDelete || otherMatchToDelete

    }

    /**
     * Returns the number of matches stored in the database.
     *
     * @return Int – Number of matches stored in DB.
     */
    fun getMatchesCount(): Int {
        val kurzInfos = DBManager.instance().getMatchesKurzInfo(-1)
        return kurzInfos.size
    }

    /**
     * Returns the number of HRF entries stored in the database.
     *
     * @return Int – Number of HRF records in DB.
     */
    fun getHrfCount(): Int {
        val allHrfs = DBManager.instance().loadAllHRFs(true)
        return allHrfs.size
    }

    private fun reInitHO() {
       RefreshManager.instance().doReInit()
    }
}
//
//fun main() {
//    HOVerwaltung.instance().loadLatestHoModel()
//    val ho = HO::class.java
//    val field = ho.getDeclaredField("versionType")
//    field.isAccessible = true
//    field.set(ho, "DEV")
//
//    val frame = JFrame()
//    frame.title = "Test"
//    frame.defaultCloseOperation = JFrame.EXIT_ON_CLOSE
//
//    DBCleanupDialog(frame, DBCleanupTool())
//}